### DISCUSSION TITLE
Basic frequency rule is not working properly with timeframe

### DISCUSSION THREAD

**Discussion by rschirin** [jertel/elastalert2#815] [Q&A] [answered]
Hey there, 
I am trying a basic frequency rule as below:

```
name: Example frequency rule via Email
type: frequency

index: log-*
num_events: 1
timeframe:
    days: 80

alert:
- "email"
email:
- "myaddress@mydomain.com"
smtp_host: "mysmtp.net"
smtp_port: 587
from_addr: "noreply@mydomain.cloud"
smtp_auth_file: "/opt/elastalert/smtp.yaml"
```

but no alert raises. If I check the documents inside elastalert_status index I will find something like:

      {
        "_index" : "elastalert_status_status",
        "_type" : "_doc",
        "_id" : "PRpkGYABVPU7w0C237Gu",
        "_score" : 1.0,
        "_source" : {
          "rule_name" : "Example frequency rule via Email",
          "endtime" : "2022-04-11T16:11:22.909285Z",
          "starttime" : "2022-04-11T16:07:44.109496Z",
          "matches" : 0,
          "hits" : 0,
          "@timestamp" : "2022-04-11T16:11:22.923959Z",
          "time_taken" : 0.014598846435546875
        }
      }

I cannot understand what is happening. Is there any error in my rule config file?

**Comment by nsano-rururu** [comment]
You can run it with the es_debug_trace option to see the query issued to elasticsearch.
https://elastalert2.readthedocs.io/en/latest/recipes/faq.html?highlight=es_debug_trace%20#my-rule-is-not-getting-any-hits

**Reply by rschirin** [reply]
Hi,
can I do using docker-compose file? In the past I got some tricky behaviors

**Reply by nsano-rururu** [reply]
I'm not interested in what you did in the past.

**Reply by rschirin** [reply]
Starting it with --verbose I get these logs:

```
Index elastalert_status already exists. Skipping index creation.
INFO:elastalert:1 rules loaded
INFO:elastalert:Starting up
INFO:elastalert:Disabled rules are: []
INFO:elastalert:Sleeping for 59.999588 seconds
INFO:elastalert:Query start and end time converting UTC to query_timezone : Europe/Rome
INFO:elastalert:Queried rule Example frequency rule via Email from 2022-04-11 16:36 UTC to 2022-04-11 16:37 UTC: 0 / 0 hits
INFO:elastalert:Ran Example frequency rule via Email from 2022-04-11 16:36 UTC to 2022-04-11 16:37 UTC: 0 query hits (0 already seen), 0 matches, 0 alerts sent
INFO:elastalert:Example frequency rule via Email range 76
INFO:elastalert:Disabled rules are: []
INFO:elastalert:Sleeping for 59.998676 seconds
INFO:elastalert:Background configuration change check run at 2022-04-11 16:38 UTC
INFO:elastalert:Background alerts thread 0 pending alerts sent at 2022-04-11 16:38 UTC
INFO:elastalert:Query start and end time converting UTC to query_timezone : Europe/Rome
INFO:elastalert:Queried rule Example frequency rule via Email from 2022-04-11 16:36 UTC to 2022-04-11 16:38 UTC: 0 / 0 hits
INFO:elastalert:Ran Example frequency rule via Email from 2022-04-11 16:36 UTC to 2022-04-11 16:38 UTC: 0 query hits (0 already seen), 0 matches, 0 alerts sent
INFO:elastalert:Example frequency rule via Email range 140
```

then, I used `--es-debug-trace` and this is the result:

```
curl -H 'Content-Type: application/json' -XPOST 'http://localhost:9200/seclog-*/_search?pretty&_source_includes=%2A%2C%40timestamp&ignore_unavailable=true&scroll=30s&size=10000' -d '{
  "query": {
    "bool": {
      "filter": {
        "bool": {
          "must": [
            {
              "range": {
                "@timestamp": {
                  "gt": "2022-04-11T18:39:35.464267+02:00",
                  "lte": "2022-04-11T18:42:06.573748+02:00"
                }
              }
     
```     
basically it is ignoring the specified timeframe

**Reply by nsano-rururu** [reply]
so what? .. Find out for yourself

**Reply by rschirin** [reply]
Hey man, what kind of answers are your? I was here just to ask a little help from collaborators or community, not to hear to me "do your homework again". 
I posted my configuration rule, I posted the logs (using your tip, thanks) so probably I am not catching the heart of matter. I supposed this was the right place to be addressed. 
So, let me stuck in reverse the doubt: is timeframe correctly used? I read the documentation and created the rule from scratch but the issue is still the same

**Reply by gminog** [reply]
> 

@rschirin you are searching via cURL for documents in `seclog-*`, while the rule searches for documents in `log-*`

**Reply by rschirin** [reply]
> > 
> 
> @rschirin you are searching via cURL for documents in `seclog-*`, while the rule searches for documents in `log-*`

`log-*` was just a typo during my copy/paste. Config rule is properly working on `seclog-*`
Anyway, unfortunately, issue is not related to queried index but `timeframe` usage.
I supposed also something about indentation but generally this will give an error.

**Comment by SysAdmLS** [comment, marked as answer]
> Hey there, I am trying a basic frequency rule as below:
> 
> ```
> name: Example frequency rule via Email
> type: frequency
> 
> index: log-*
> num_events: 1
> timeframe:
>     days: 80
> 
> alert:
> - "email"
> email:
> - "myaddress@mydomain.com"
> smtp_host: "mysmtp.net"
> smtp_port: 587
> from_addr: "noreply@mydomain.cloud"
> smtp_auth_file: "/opt/elastalert/smtp.yaml"
> ```
> 
> but no alert raises. If I check the documents inside elastalert_status index I will find something like:
> 
> ```
>   {
>     "_index" : "elastalert_status_status",
>     "_type" : "_doc",
>     "_id" : "PRpkGYABVPU7w0C237Gu",
>     "_score" : 1.0,
>     "_source" : {
>       "rule_name" : "Example frequency rule via Email",
>       "endtime" : "2022-04-11T16:11:22.909285Z",
>       "starttime" : "2022-04-11T16:07:44.109496Z",
>       "matches" : 0,
>       "hits" : 0,
>       "@timestamp" : "2022-04-11T16:11:22.923959Z",
>       "time_taken" : 0.014598846435546875
>     }
>   }
> ```
> 
> I cannot understand what is happening. Is there any error in my rule config file?

yes there is, you did not define a filter to match something and thus creating an event.

Check out the example frequency rule : https://github.com/jertel/elastalert2/blob/master/examples/rules/example_frequency.yaml

**Reply by rschirin** [reply]
Ok, now I used the example frequency rule defined in the official project (using also the required filter field) but the result is the same.
config file:

```
name: Example frequency rule
type: frequency
index: seclog-*
num_events: 50
timeframe:
  hours: 4
filter:
- term:
    severity: "ERROR"
```

output log
```
curl -H 'Content-Type: application/json' -XPOST 'http://localhost:9200/seclog-*/_search?pretty&_source_includes=%2A%2C%40timestamp&ignore_unavailable=true&scroll=30s&size=10000' -d '{
  "query": {
    "bool": {
      "filter": {
        "bool": {
          "must": [
            {
              "range": {
                "@timestamp": {
                  "gt": "2022-04-12T10:19:29.812931+02:00",
                  "lte": "2022-04-12T10:34:29.812931+02:00"
                }
              }
            },
            {
              "term": {
                "severity": "ERROR"
              }
            }
          ]
        }
      }
    }
  },
  "sort": [
    {
      "@timestamp": {
        "order": "asc"
      }
    }
  ]
}'
```

It doesn't make sense to me this behavior.

**Reply by SysAdmLS** [reply]
Try testing that rule as explained here https://elastalert2.readthedocs.io/en/latest/ruletypes.html#testing-your-rule and make sure your filter is correct (the "" around the field key (severity) could be wrong)

See if it gets any hits. If not post example data here that you are trying to generate an alert for.

**Reply by rschirin** [reply]
I tested and I can say that filter is correct. Query is correct but, again, it is not using the correct timeframe. I want to check the last 4 hours (or 10 days, it doesn't matter) while it is running a query on the last 5 minutes.

**Reply by SysAdmLS** [reply]
> I tested and I can say that filter is correct. Query is correct but, again, it is not using the correct timeframe. I want to check the last 4 hours (or 10 days, it doesn't matter) while it is running a query on the last 5 minutes.

I believe thats how elastalert is intended to work, it stores the previous found matches and the last time it ran and then simply does addition with the new found matches(could be wrong, thats how ive understand from docs). You can use scan_entire_timeframe: True in your rule to force the query to always scan the whole timeframe you defined.
